### PR TITLE
Fix mixed route for action error message

### PR DESCRIPTION
**Pull request by kichalla**
@rynowak 

https://github.com/aspnet/Entropy/pull/106

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs [hunk 1]
```diff
@@ -727,8 +727,16 @@ private static bool IsAttributeRoutedAction(ControllerActionDescriptor actionDes
             {
                 var routeTemplate = action.AttributeRouteInfo?.Template ?? nullTemplate;
 
-                var verbs = action.ActionConstraints.OfType<HttpMethodActionConstraint>().FirstOrDefault()?.HttpMethods;
-                var formattedVerbs = string.Join(", ", verbs.OrderBy(v => v, StringComparer.Ordinal));
+                var verbs = action.ActionConstraints?
+                    .OfType<HttpMethodActionConstraint>()
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs]: Never split `?.`

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs [hunk 2]
```diff
@@ -727,8 +727,16 @@ private static bool IsAttributeRoutedAction(ControllerActionDescriptor actionDes
...
+                    .FirstOrDefault()?
+                    .HttpMethods;
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs]: Was the previous statement actually incorrect? I.e. was `ActionConstraints` really `null` in the failing case (or was it just that no defined constraints had type `HttpMethodActionConstraint`?

**Comment by dougbu** [comment]
Suggest adding a unit test.

**Comment by dougbu** [comment]
Are CI builds failing for known reasons?

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs (hunk 1 above)
Comment (Eilon) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs]: Curious - what would you do instead? We often like the `.` to start the next line, so what do you do here?

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs]: It's a single operator so I'd start the line w/ `?.`.

Comment (Eilon) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs]: Ah, good point.

**Comment by kichalla** [comment]
:up: :date:

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs (hunk 2 above)
Comment (kichalla) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs]: Yes `ActionConstraints` was null.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs [hunk 3]
```diff
@@ -721,8 +721,15 @@ private static bool IsAttributeRoutedAction(ControllerActionDescriptor actionDes
             {
                 var routeTemplate = action.AttributeRouteInfo?.Template ?? nullTemplate;
 
-                var verbs = action.ActionConstraints.OfType<HttpMethodActionConstraint>().FirstOrDefault()?.HttpMethods;
-                var formattedVerbs = string.Join(", ", verbs.OrderBy(v => v, StringComparer.Ordinal));
+                var verbs = action.ActionConstraints?.OfType<HttpMethodActionConstraint>()
+                    .FirstOrDefault()?
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs]: Don't split the `?.` across lines, it's a single operator.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs [hunk 4]
```diff
@@ -721,8 +721,15 @@ private static bool IsAttributeRoutedAction(ControllerActionDescriptor actionDes
...
+                    .HttpMethods;
+
+                var formattedVerbs = string.Empty;
+                if (verbs != null)
+                {
+                    formattedVerbs = string.Join(", ", verbs.OrderBy(v => v, StringComparer.Ordinal));
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs]: This should really be Invariant

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/ControllerActionDescriptorProviderTests.cs [hunk 5]
```diff
@@ -775,6 +773,47 @@ public void AttributeRouting_ThrowsIfAttributeRoutedAndNonAttributedActions_OnTh
         }
 
         [Fact]
+        public void AttributeRouting_ThrowsIfAttributeRoutedAndNonAttributedActions_OnTheSameMethod_UsingCustomConvention()
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/ControllerActionDescriptorProviderTests.cs]: Leave a comment here explaining what this is all about

**Comment by rynowak** [comment]
:shipit:

**Comment by dougbu** [comment]
:shipit:

**Pull request merged**